graphene_federation/schema.py

```diff
@@ -0,0 +1,200 @@
+from typing import Collection, Type, Union
+from typing import Optional
+
+from graphene import ObjectType, PageInfo
+from graphene_directives import (
+    SchemaDirective,
+    build_schema as build_directive_schema,
+    directive_decorator,
+)
+from graphene_directives.schema import Schema
+
+from .apollo_versions import (
+    FederationVersion,
+    STABLE_VERSION,
+    get_directive_from_name,
+    get_directives_based_on_version,
+)
+from .apollo_versions.v2_1 import compose_directive as ComposeDirective
+from .composable_directive import ComposableDirective
+from .entity import get_entity_query
+from .schema_directives import compose_directive, link_directive
+from .service import get_service_query
+
+
+def _get_federation_query(
+    schema: Schema, query_cls: Optional[ObjectType] = None
+) -> Type[ObjectType]:
+    """
+    Add Federation required _service and _entities to Query(ObjectType)
+    """
+    type_name = "Query"
+    bases = [get_service_query(schema)]
+    entity_cls = get_entity_query(schema)
+    if entity_cls:
+        bases.append(entity_cls)
+    if query_cls is not None:
+        type_name = query_cls.__name__
+        bases.append(query_cls)
+    federated_query_cls = type(type_name, tuple(bases), {})
+    return federated_query_cls  # noqa
+
+
+def _add_sharable_to_page_info_type(
+    schema: Schema,
+    federation_version: FederationVersion,
+    types: list[Union[ObjectType, Type[ObjectType]]],
+):
+    """
+    Add @sharable directive to PageInfo type
+    """
+    if PageInfo.__name__ in schema.graphql_schema.type_map:
+        try:
+            # PageInfo needs @sharable directive
+            sharable = get_directive_from_name("shareable", federation_version)
+            types.append(
+                directive_decorator(target_directive=sharable)(field=None)(PageInfo)
```

Comment:
In a multi-thread environment, this function could be called multiple times. If `PageInfo` is already marked as `@shareable`, then an exception is raised:

<details>

<summary>Large stack trace from Django startup</summary>

```
Exception in thread django-main-thread:
Traceback (most recent call last):
  File "/usr/local/lib/python3.11/threading.py", line 1045, in _bootstrap_inner
    self.run()
  File "/usr/local/lib/python3.11/threading.py", line 982, in run
    self._target(*self._args, **self._kwargs)
  File "/home/django/.venv/lib/python3.11/site-packages/django/utils/autoreload.py", line 64, in wrapper
    fn(*args, **kwargs)
  File "/home/django/.venv/lib/python3.11/site-packages/django/core/management/commands/runserver.py", line 133, in inner_run
    self.check(display_num_errors=True)
  File "/home/django/.venv/lib/python3.11/site-packages/django/core/management/base.py", line 485, in check
    all_issues = checks.run_checks(
                 ^^^^^^^^^^^^^^^^^^
  File "/home/django/.venv/lib/python3.11/site-packages/django/core/checks/registry.py", line 88, in run_checks
    new_errors = check(app_configs=app_configs, databases=databases)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/django/.venv/lib/python3.11/site-packages/django/core/checks/urls.py", line 14, in check_url_config
    return check_resolver(resolver)
           ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/django/.venv/lib/python3.11/site-packages/django/core/checks/urls.py", line 24, in check_resolver
    return check_method()
           ^^^^^^^^^^^^^^
  File "/home/django/.venv/lib/python3.11/site-packages/django/urls/resolvers.py", line 494, in check
    for pattern in self.url_patterns:
                   ^^^^^^^^^^^^^^^^^
  File "/home/django/.venv/lib/python3.11/site-packages/django/utils/functional.py", line 57, in __get__
    res = instance.__dict__[self.name] = self.func(instance)
                                         ^^^^^^^^^^^^^^^^^^^
  File "/home/django/.venv/lib/python3.11/site-packages/django/urls/resolvers.py", line 715, in url_patterns
    patterns = getattr(self.urlconf_module, "urlpatterns", self.urlconf_module)
                       ^^^^^^^^^^^^^^^^^^^
  File "/home/django/.venv/lib/python3.11/site-packages/django/utils/functional.py", line 57, in __get__
    res = instance.__dict__[self.name] = self.func(instance)
                                         ^^^^^^^^^^^^^^^^^^^
  File "/home/django/.venv/lib/python3.11/site-packages/django/urls/resolvers.py", line 708, in urlconf_module
    return import_module(self.urlconf_name)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<frozen importlib._bootstrap>", line 1204, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1176, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1147, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 690, in _load_unlocked
  File "/home/django/.venv/lib/python3.11/site-packages/ddtrace/internal/module.py", line 220, in _exec_module
    self.loader.exec_module(module)
  File "<frozen importlib._bootstrap_external>", line 940, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "/home/django/app/partner_library/urls.py", line 2, in <module>
    from partner_library.graphql_schema import schema
  File "<frozen importlib._bootstrap>", line 1176, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1147, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 690, in _load_unlocked
  File "/home/django/.venv/lib/python3.11/site-packages/ddtrace/internal/module.py", line 220, in _exec_module
    self.loader.exec_module(module)
  File "/home/django/app/partner_library/graphql_schema.py", line 1689, in <module>
    schema = build_schema(
             ^^^^^^^^^^^^^
  File "/home/django/.venv/lib/python3.11/site-packages/graphene_federation/schema.py", line 120, in build_schema
    _add_sharable_to_page_info_type(
  File "/home/django/.venv/lib/python3.11/site-packages/graphene_federation/schema.py", line 56, in _add_sharable_to_page_info_type
    directive_decorator(target_directive=sharable)(field=None)(PageInfo)
  File "/home/django/.venv/lib/python3.11/site-packages/graphene_directives/directive.py", line 177, in decorator
    set_attribute_value(
  File "/home/django/.venv/lib/python3.11/site-packages/graphene_directives/utils.py", line 64, in set_attribute_value
    raise DirectiveValidationError(
graphene_directives.exceptions.DirectiveValidationError: @shareable is got duplicate values {}, at: PageInfo
```

</details>

Maybe `_add_sharable_to_page_info_type()` should check to make sure the directive hasn't already been applied?

Reply:
Thanks. I have written a fix for this [74ce029](https://github.com/strollby/graphene-federation/pull/20/commits/74ce0290dfb4d29c5f3607e3fdefcb75f01fc52f). Let me know if it works in django

Reply:
New error from this change, looks like `directive_decorator()()(type_)` expects a class, not an instance:

```
File "/home/django/.venv/lib/python3.11/site-packages/graphene_federation/schema.py", line 121, in build_schema
    _add_sharable_to_page_info_type(
  File "/home/django/.venv/lib/python3.11/site-packages/graphene_federation/schema.py", line 57, in _add_sharable_to_page_info_type
    directive_decorator(target_directive=sharable)(field=None)(page_info)
  File "/home/django/.venv/lib/python3.11/site-packages/graphene_directives/directive.py", line 169, in decorator
    if not any(
           ^^^^
  File "/home/django/.venv/lib/python3.11/site-packages/graphene_directives/directive.py", line 170, in <genexpr>
    LOCATION_NON_FIELD_VALIDATOR[valid_type](type_)
  File "/home/django/.venv/lib/python3.11/site-packages/graphene_directives/constants.py", line 9, in <lambda>
    GrapheneDirectiveLocation.OBJECT: lambda t: issubclass(t, graphene.ObjectType),
                                                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
TypeError: issubclass() arg 1 must be a class
```

Reply:
Have fixed it now. Can you recheck on django

Reply:
That fixed this issue, thanks!

Reply:
Awesome.